src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java

```diff
@@ -202,6 +205,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
         } catch (Exception e) {
             log.error("{}{}", PATCH_CLOSE_PAYMENT_ERROR, requestId, e);
             log.info("Refunding payment with requestId: {}", requestId);
+            if (executeOrderStatus(entity, pgsRequest).equals(OK)) {
+                executeRevert(entity, pgsRequest);
+            }
+        }
+    }
+
+    private RefundOutcome executeOrderStatus(PaymentRequestEntity entity, StepZeroRequest stepZeroRequest) throws IOException {
+        log.info("Calling VPOS - OrderStatus - for requestId: " + entity.getGuid());
+        Map<String, String> params = vPosRequestUtils.buildOrderStatusParams(stepZeroRequest);
+        HttpClientResponse clientResponse = callVPos(params);
+        VposOrderStatusResponse response = vPosResponseUtils.buildOrderStatusResponse(clientResponse.getEntity());
+        return computeOrderStatusResultCode(response, entity);
+    }
+
+    private RefundOutcome computeOrderStatusResultCode(VposOrderStatusResponse response, PaymentRequestEntity entity) {
+        String resultCode = response.getResultCode();
+        if (resultCode.equals(RESULT_CODE_AUTHORIZED)) {
+            return OK;
+        } else {
+            entity.setErrorMessage("Error during orderStatus");
+            paymentRequestRepository.save(entity);
+            return KO;
+        }
+    }
+
+    private void executeRevert(PaymentRequestEntity entity, StepZeroRequest pgsRequest) {
+        try {
+            log.info("Calling VPOS - Revert - for requestId: " + entity.getGuid());
+            Map<String, String> params = vPosRequestUtils.buildRevertRequestParams(pgsRequest, entity.getCorrelationId());
+            HttpClientResponse clientResponse = callVPos(params);
+            AuthResponse response = vPosResponseUtils.buildAuthResponse(clientResponse.getEntity());
+            vPosResponseUtils.validateResponseMac(response.getTimestamp(), response.getResultCode(), response.getResultMac(), pgsRequest);
+            checkRevertResultCode(response, entity);
+        } catch (Exception e) {
+            log.error(GENERIC_REFUND_ERROR_MSG + entity.getIdTransaction() + " cause: " + e.getCause() + " - " + e.getMessage(), e);
+        }
+    }
+
+    private void checkRevertResultCode(AuthResponse response, PaymentRequestEntity entity) {
+        String resultCode = response.getResultCode();
+        if (resultCode.equals(RESULT_CODE_AUTHORIZED)) {
+            entity.setStatus(CANCELLED.name());
+            entity.setIsRefunded(true);
+            paymentRequestRepository.save(entity);
+        } else {
+            entity.setErrorMessage("Error during Revert");
+            entity.setIsRefunded(false);
         }
+        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid() + " - resultCode: " + resultCode);
     }
```

Comment:
This function has a duplicate in `CcResumeStep1Service.java`

Reply:
https://github.com/pagopa/pagopa-payment-transactions-gateway/commit/7035740f5c2b7c23029194ae1fa443550e0e1050 I did a refactoring